### PR TITLE
feat: Refactor notifications and make them injectable

### PR DESCRIPTION
**Pull request by gmalette**
### Description

First I think it's worth giving context to what my goal is. I would like to create a program that wraps streetmerchant and offers another dashboard. I would like for it to run in the same process and avoid adding infrastructure if at all possible.

This PR refactors the notifications to make both failures and success notifications. It also makes the notification function injectable, but the `main` function uses the existing one.

I also refactored to make the logging of failures as part of normal notifications. This isn't strictly necessary but I thought it was nice to harmonize both and show that logging is just one way to handle poll events.

This PR, along with https://github.com/jef/streetmerchant/compare/main...gmalette:gm/packaging?expand=1, allow me to use streetmerchant and inject notification handling.

```ts
import {Browser, launch} from 'puppeteer';
import {config} from 'streetmerchant/build/src/config';
import {getSleepTime} from 'streetmerchant/build/src/util';
import {storeList} from 'streetmerchant/build/src/store/model';
import {tryLookupAndLoop} from 'streetmerchant/build/src/store';
import {LinkPollEvent} from 'streetmerchant/build/src/notification/link_poll_event';

const sendNotification = (event: LinkPollEvent): void => {
  console.log(event);
};

async function main() {
  const args: string[] = [];
  const browser: Browser = await launch({
    args,
    defaultViewport: {
      height: config.page.height,
      width: config.page.width,
    },
    headless: config.browser.isHeadless,
  });

  config.browser.userAgent = await browser.userAgent();

  for (const store of storeList.values()) {
    if (store.setupAction !== undefined) {
      store.setupAction(browser);
    }

    setTimeout(tryLookupAndLoop, getSleepTime(store), browser, store, sendNotification);
  }
}

void main();
```

### Testing

Ran the program, worked as normal for "out of stock" and other errors.
I was not able to test for "in stock" items.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -76,7 +77,13 @@ async function main() {
       store.setupAction(browser);
     }
 
-    setTimeout(tryLookupAndLoop, getSleepTime(store), browser, store);
+    setTimeout(
+      tryLookupAndLoop,
+      getSleepTime(store),
+      browser,
+      store,
+      sendNotification
```
Comment (gmalette) [inline src/index.ts:85]: In the `main` function we inject the existing `sendNotification` function

File: src/notification/link_poll_event.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  | MaxPriceExceeded
+  | OutOfStock
+  | BannedSeller
+);
+type LinkPollSuccess = {result: 'in_stock'; url: string};
+
+export type LinkPollEvent = (LinkPollError | LinkPollSuccess) & {
+  link: Link;
+  store: Store;
+};
```
Comment (gmalette) [inline src/notification/link_poll_event.ts:23]: The `LinkPollEvent` could be made shorter but I liked to express it in a way that prevents representing illegal states.

File: src/notification/logstream.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      case 'max_price':
+        logger.info(Print.maxPrice(link, store, pollEvent.maxPrice, true));
+        return;
+      case 'request_failed':
+        return;
+      default:
+        assertNever(pollEvent);
+    }
+  }
+}
```
Comment (gmalette) [inline src/notification/logstream.ts:32]: All the logging w.r.t polling had been moved here. Some logging still happends in `lookup.ts` but looked ancillary, except for what's mentioned in the comments on `notifications`.

File: src/store/lookup.ts [hunk 4]
```diff
@@ -151,7 +153,11 @@ async function handleAdBlock(request: HTTPRequest, adBlockRequestHandler: any) {
  * @param browser Puppeteer browser.
  * @param store Vendor of graphics cards.
  */
-async function lookup(browser: Browser, store: Store) {
+async function lookup(
+  browser: Browser,
+  store: Store,
+  sendNotification: SendNotification
```
Comment (gmalette) [inline src/store/lookup.ts:159]: Most of the added code in this file is passing the `sendNotification` to the functions that need it. 

The rest is sending notifications in lieu of logging

File: src/notification/logstream.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        logger.warn(Print.bannedSeller(link, store, true));
+        return;
+      case 'out_of_stock':
+        logger.info(Print.outOfStock(link, store, true));
+        return;
+      case 'max_price':
+        logger.info(Print.maxPrice(link, store, pollEvent.maxPrice, true));
+        return;
+      case 'request_failed':
+        return;
```
Comment (gmalette) [inline src/notification/logstream.ts:27]: I think this case logs too but I haven't found where it happens.

**Comment by jef** [comment]
I'll be looking at this and #2038 very shortly.

I appreciate your contribution!

File: src/notification/notification.ts [hunk 6]
```diff
@@ -17,8 +17,57 @@ import {sendTwitchMessage} from './twitch';
...
+export type SendNotification = (pollEvent: LinkPollEvent) => void;
+
+function assertNever(x: never): never {
+  throw new Error('Unexpected object: ' + x);
+}
+
+export function sendNotification(pollEvent: LinkPollEvent) {
+  const {link, store} = pollEvent;
+  if (pollEvent.result === 'failure') {
+    switch (pollEvent.failureReason) {
```
Comment (jef) [inline src/notification/notification.ts]:
```suggestion
  const {failureReason, link, result, store} = pollEvent;
  if (result === 'failure') {
    switch (failureReason) {
```

Might as well yank these out as well.

**Review by jef — COMMENTED** [review]
Looks good! I'm not 100% certain if folks want emails and such _everytime_ captcha, out of stock, and etc show up.

I do understand the concept, but I'm not certain if we want this behavior.

That being said, if you make it configurable, I'll be more than happy to pull it in!

**Comment by gmalette** [comment]
@jef I'm not sure I understand what kind of changes you're asking for. The default injected `sendNotification` will stop after [sending the logstream event unless it's a success](https://github.com/jef/streetmerchant/pull/2034/files#diff-94009a56ba9cbb3115976ea275dab1109adc6a18336aa3b61a987eeda481a0edR24-R26), meaning that the existing behaviour should remain unaltered.

**Comment by jef** [comment]
Sorry for the late response @gmalette, this have been pretty busy on my end.

Let me take another look at this shortly and we can get this in.

**Comment by gmalette** [comment]
LMK if you'd like me to reopen this, it would probably help with https://github.com/jef/streetmerchant/issues/2337, especially now that we have https://github.com/jef/streetmerchant/pull/2066

**Pull request closed without merging**